design/task-management-api.md

```diff
@@ -0,0 +1,346 @@
+# Task management API: requirements
+
+Multiple subsystems require reliable long-lived distributed operations.  We shall support them
+with a task queue subsystem.  Concepts of the API are intended to parallel concepts in the
+[Celery][celery-api] and/or [Machinery][machinery-api] APIs.  We do not use either of them to
+reduce the number of required dependencies.  A future version may support external queues to
+achieve better performance or robustness, at the price of increased ops or cost.  These will
+likely be queues such as Kafka or SQS rather than full Celery or Machinery.
+
+In practice this generally means providing APIs similar to those of Machinery (which is more
+Go-like than Celery) for constructing task flows and for registering workers.
+
+In particular:
+1. We provide similar concepts for building task flows as do existing
+   task queues.
+1. We use similar terminology.
+1. We do *not* require the entire API of an existing task queue.
+1. We do *not* use the verbs or API calls of an existing task queue.
+
+This API definition comes with implementation sketches for how we to use these APIs to implement
+the branch export and concurrent dedupe "user" stories.
+
+## API
+
+### Concepts
+
+#### Tasks
+
+A task is the basic atom of task management.  It represents a single unit of work to
+perform, and can succeed or fail.  Tasks may be retried on failure, so _executing a task
+must be idempotent_.
+
+Tasks include these attributes:
+- `Id`: a unique identifier for the task.  Use a known-unique substring in the identifier
+  (e.g. a UUID or [nanoid][nanoid]) to avoid collisions, or a well-known identifier to ensure
+  only one task of a type can exist.
+- `Action`: the type of action to perform for this task.  Workers pick tasks to perform and the
```

Comment:
How do we define what an action does? Are the actions go methods that are registered on startup? or something else?

Reply:
My understanding is that a task runs some custom code like copying an object etc. I do not see how this code is written and activated. This is a pretty heavy subject as it bridges between the infrastructure queuing mechanism, and the application needs

Reply:
This is one way to do it but is certainly not the _only_ way.  We can provide a specific higher-level API to do this.  The reason not to start by providing it is that it has _so many options_, and none of them are important to functionality.  These options will include:
* What to do when no tasks are available
  For some job types it makes sense to terminate the loop once there are no more available tasks.  For others it makes sense to back off (with various strategies) and retry.
* Worker function signature
* Serialization of body
* Action deadline specification
  - Whether it auto-renews, and if so -- on which context
* Cancellation support
* How to handle asynchronous workers
  Easiest API is to have a "synchronous" worker that calls additional goroutines to do its work.  But the API here does not force even that, e.g. you can pass a task ID and a performance token to another goroutine that returns the task

I'd like to revisit this later.